From 852e228e9d85bd1c5a3a77b8a867b38ae41c43a5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Jirman?= <megi@xff.cz>
Date: Thu, 27 Jan 2022 00:45:19 +0100
Subject: [PATCH 396/456] power: supply: ip5xxx-power: Fix current/voltage
 reporting

Charging was reported as a negative current and had invalid value.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/power/supply/ip5xxx_power.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/drivers/power/supply/ip5xxx_power.c b/drivers/power/supply/ip5xxx_power.c
index ae5d3167594e..4e4e8f2144bc 100644
--- a/drivers/power/supply/ip5xxx_power.c
+++ b/drivers/power/supply/ip5xxx_power.c
@@ -293,7 +293,10 @@ static int ip5xxx_charger_read_adc(struct ip5xxx *ip5xxx,
 	if (ret)
 		return ret;
 
-	*val = sign_extend32(hi << 8 | lo, 13);
+	if (hi & 0x20)
+		*val = -(int)((~lo & 0xff) + ((~hi & 0x1f) << 8) + 1);
+	else
+		*val = hi << 8 | lo;
 
 	return 0;
 }
@@ -322,22 +325,28 @@ static int ip5xxx_charger_get_property(struct power_supply *psy,
 	case POWER_SUPPLY_PROP_VOLTAGE_NOW:
 		ret = ip5xxx_charger_read_adc(ip5xxx, IP5XXX_BATVADC_DAT0,
 					      IP5XXX_BATVADC_DAT1, &raw);
+		if (ret)
+			return ret;
 
-		val->intval = 2600000 + DIV_ROUND_CLOSEST(raw * 26855, 100);
+		val->intval = (2600 + DIV_ROUND_CLOSEST(raw * 1000, 3724)) * 1000;
 		return 0;
 
 	case POWER_SUPPLY_PROP_VOLTAGE_OCV:
 		ret = ip5xxx_charger_read_adc(ip5xxx, IP5XXX_BATOCV_DAT0,
 					      IP5XXX_BATOCV_DAT1, &raw);
+		if (ret)
+			return ret;
 
-		val->intval = 2600000 + DIV_ROUND_CLOSEST(raw * 26855, 100);
+		val->intval = (2600 + DIV_ROUND_CLOSEST(raw * 1000, 3724)) * 1000;
 		return 0;
 
 	case POWER_SUPPLY_PROP_CURRENT_NOW:
 		ret = ip5xxx_charger_read_adc(ip5xxx, IP5XXX_BATIADC_DAT0,
 					      IP5XXX_BATIADC_DAT1, &raw);
+		if (ret)
+			return ret;
 
-		val->intval = DIV_ROUND_CLOSEST(raw * 745985, 1000);
+		val->intval = DIV_ROUND_CLOSEST(raw * 1000, 1341) * 1000;
 		return 0;
 
 	case POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT:
-- 
2.34.1

